### PR TITLE
test: add shadow traffic tests

### PR DESCRIPTION
**Pull request by szuecs**
test: add shadow traffic tests

### CODE REVIEW

File: proxy/shadow_test.go [hunk 1]
```diff
@@ -0,0 +1,405 @@
...
+
+	ts := stdlibhttptest.NewServer(pr)
+	defer ts.Close()
+
+	reqURL, err := url.Parse(ts.URL)
+	if err != nil {
+		t.Fatalf("Failed to parse url %s: %v", ts.URL, err)
+	}
+
+	client := http.DefaultClient
```
Comment (AlexanderYastrebov) [inline proxy/shadow_test.go:98]: Why not use ts.Client() here?

File: proxy/shadow_test.go [hunk 2]
```diff
@@ -0,0 +1,405 @@
...
+	})
+	defer pr.Close()
+
+	ts := stdlibhttptest.NewServer(pr)
+	defer ts.Close()
+
+	reqURL, err := url.Parse(ts.URL)
+	if err != nil {
+		t.Fatalf("Failed to parse url %s: %v", ts.URL, err)
+	}
```
Comment (AlexanderYastrebov) [inline proxy/shadow_test.go:96]: I think we should just use `ts.URL` in the request there is no need to parse it

File: proxy/shadow_test.go [hunk 3]
```diff
@@ -0,0 +1,405 @@
...
+	fr.Register(scheduler.NewFifo())
+
+	doc := fmt.Sprintf(`
+main: PathSubtree("/")
+-> "%s";
+split: PathSubtree("/") && Traffic(0.5)
+-> teeLoopback("test")
+-> "%s";
+shadow: PathSubtree("/") && Tee("test") && True()
+-> "%s";
```
Comment (AlexanderYastrebov) [inline proxy/shadow_test.go:61]: Maybe if we indent inside of string block it would look nicer?

File: proxy/shadow_test.go (hunk 1 above)
Comment (szuecs) [reply proxy/shadow_test.go:98]: done

File: proxy/shadow_test.go (hunk 2 above)
Comment (szuecs) [reply proxy/shadow_test.go:96]: done

File: proxy/shadow_test.go (hunk 3 above)
Comment (szuecs) [reply proxy/shadow_test.go:61]: done

**Comment by szuecs** [comment]
:+1:

File: proxy/shadow_test.go [hunk 4]
```diff
@@ -0,0 +1,398 @@
...
+					return fmt.Errorf("fifo full %d", statusFifoFull)
+				}
+				if statusFifoTimeout == 0 {
+					return fmt.Errorf("fifo timeout %d", statusFifoTimeout)
+				}
+				if statusFifoErr != 0 {
+					return fmt.Errorf("fifo err %d", statusFifoErr)
+				}
+				return nil
+			},
```
Comment (AlexanderYastrebov) [inline proxy/shadow_test.go:232]: This is quite hard to read when placed in the table test. Alternatively you could have a test helper method that setups proxy and runs load and then use it in a series of nested t.Runs like
```go
t.Run("foo", func(t *testing.T) {
	va := runTest(`
	  main: PathSubtree("/")
            -> fifo(2, 5, "200ms")
            -> "$backendURL";
          split: PathSubtree("/") && Traffic(1.0)
            -> fifo(2, 5, "200ms")
            -> teeLoopback("test")
            -> "$backendURL";
          shadow: PathSubtree("/") && Tee("test") && True()
            -> fifo(2, 5, "200ms")
            -> "$shadowURL";
	`)
	...
})
```

**Comment by AlexanderYastrebov** [comment]
:+1:

**Pull request merged**